File: Python/0051-0100/53. Maximum Subarray.py
# https://leetcode.com/problems/maximum-subarray/description/

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        result = -922337203685477580 # possible minimum value
        sum = 0 # subarraySum

        for num in nums:
            sum += num # add the current element to subarray sum
            # update result
            result = max(result, sum)

            if sum < 0:
                sum = 0
        
        return result;

File: Java/0051-0100/53. Maximum Subarray.java
// https://leetcode.com/problems/maximum-subarray/description/

class Solution {
    public int maxSubArray(int[] nums) {
        int max = Integer.MIN_VALUE;
        int sum = 0; // subarray Sum

        for(int num : nums){
            sum += num; // ADD CURRENT ELEMENT TO SUBARRAY SUM
            max = Math.max(max, sum); // update the result
            // check if sum goes below 0
            if(sum < 0){
                sum = 0;
            }
        }
        return max;
    }
}
